packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts

```diff
@@ -270,19 +271,59 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
-            const gcDetailsBlob = snapshotTree.blobs[gcBlobKey];
-            if (gcDetailsBlob !== undefined) {
-                const gcDetails = await readAndParseBlob<IGarbageCollectionDetailsBase>(gcDetailsBlob);
+            // Load the base GC details before proceeding because if that happens later it can overwrite the GC details
+            // written by the following code.
+            await this.loadBaseGCDetails();
+            /**
+             * GC data is written at root of the snapshot tree under "gc" sub-tree. This data needs to be propagated to
+             * all the nodes in the container.
+             * The root summarizer node reads the GC data from the "gc" sub-tree, runs GC on it to get used routes in
+             * the container and updates its GC data and referenced used routes. It then gets the GC data and used
+             * routes of all its children and adds it to their snapshot tree.
+             * All the other nodes gets the GC data and used routes from their snapshot tree and updates their state.
+             * They get the GC data and used routes of their children and add it to their snapshot tree and so on.
+             */
+            let gcDetails: IGarbageCollectionDetailsBase = { gcData: { gcNodes: {} } };
+            const gcSnapshotTree = snapshotTree.trees[gcTreeKey];
+            if (gcSnapshotTree !== undefined) {
+                // This is the root summarizer node. It reads GC data from the root and processes it as explained above.
+                const gcSnapshotData = await getGCDataFromSnapshot(gcSnapshotTree, readAndParseBlob);
 
                 // Possible re-entrancy. If we have already seen a summary later than this one, ignore it.
                 if (this.referenceSequenceNumber >= referenceSequenceNumber) {
                     return;
                 }
 
-                this.referenceUsedRoutes = gcDetails.usedRoutes;
+                const gcNodes: { [id: string]: string[]; } = {};
+                for (const [nodeId, nodeData] of Object.entries(gcSnapshotData.gcState.gcNodes)) {
+                    gcNodes[nodeId] = Array.from(nodeData.outboundRoutes);
+                }
+                // Run GC on the nodes in the snapshot to get the used routes for each node in the container.
+                const usedRoutes = runGarbageCollection(gcNodes, ["/"]).referencedNodeIds;
+                gcDetails = { gcData: { gcNodes }, usedRoutes };
+            } else {
+                // Non root summarizer nodes. They get their GC data and used routes from their snapshot tree.
+                const gcDetailsBlob = snapshotTree.blobs[gcTreeKey];
+                gcDetails = JSON.parse(gcDetailsBlob) as IGarbageCollectionDetailsBase;
+
+                // Possible re-entrancy. If we have already seen a summary later than this one, ignore it.
+                if (this.referenceSequenceNumber >= referenceSequenceNumber) {
+                    return;
+                }
```

Comment:
Can we move this if statement to the top of this function?

Reply:
Something like
`if(this.referenceSequenceNumber >= referenceSequenceNumber || this.gcDisabled) { return; }`

Reply:
I moved it to to top of the if (!this.gcDisabled) block. I could combine it with this.gcDisabled but I think it's easier to reason with this way (base summarizer node has similar checks)